File: zipline-kotlin-plugin-tests/src/test/java/app/cash/zipline/kotlin/ZiplineTestInternals.java
/*
 * Copyright (C) 2021 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package app.cash.zipline.kotlin;

import app.cash.zipline.ZiplineFunction;
import app.cash.zipline.ZiplineScope;
import app.cash.zipline.internal.bridge.Endpoint;
import app.cash.zipline.internal.bridge.OutboundCallHandler;
import app.cash.zipline.internal.bridge.OutboundService;
import app.cash.zipline.internal.bridge.ReturningZiplineFunction;
import app.cash.zipline.internal.bridge.ZiplineServiceAdapter;
import app.cash.zipline.testing.EchoRequest;
import app.cash.zipline.testing.EchoResponse;
import app.cash.zipline.testing.EchoService;
import app.cash.zipline.testing.EchoZiplineService;
import app.cash.zipline.testing.GenericEchoService;
import app.cash.zipline.testing.KotlinSerializersKt;
import app.cash.zipline.testing.SignatureHashKt;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import kotlin.Pair;
import kotlin.PublishedApi;
import kotlin.coroutines.EmptyCoroutineContext;
import kotlin.jvm.JvmClassMappingKt;
import kotlin.reflect.KType;
import kotlin.reflect.KTypeProjection;
import kotlin.reflect.full.KClassifiers;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.CoroutineScopeKt;
import kotlinx.serialization.KSerializer;
import kotlinx.serialization.SerializersKt;
import kotlinx.serialization.modules.SerializersModule;
import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;

import androidx.annotation.NonNull;

/**
 * Call these {@link PublishedApi} internal APIs from Java rather than from Kotlin to hack around
 * visibility of internal APIs.
 */
@SuppressWarnings("KotlinInternalInJava")
public final class ZiplineTestInternals {
  private static final KType echoResponseKt = KClassifiers.createType(
      JvmClassMappingKt.getKotlinClass(EchoResponse.class), emptyList(), false, emptyList());
  private static final KType echoRequestKt = KClassifiers.createType(
      JvmClassMappingKt.getKotlinClass(EchoRequest.class), emptyList(), false, emptyList());
  private static final KType stringKt = KClassifiers.createType(
      JvmClassMappingKt.getKotlinClass(String.class), emptyList(), false, emptyList());
  private static final KType listOfStringKt = KClassifiers.createType(
      JvmClassMappingKt.getKotlinClass(List.class),
      singletonList(KTypeProjection.invariant(stringKt)), false, emptyList());

  public static Pair<Endpoint, Endpoint> newEndpointPair() {
    CoroutineScope scope = CoroutineScopeKt.CoroutineScope(EmptyCoroutineContext.INSTANCE);
    return app.cash.zipline.testing.EndpointsKt.newEndpointPair(
      scope, KotlinSerializersKt.getKotlinBuiltInSerializersModule());
  }

  /** Simulate generated code for outbound calls. */
  public static EchoService takeEchoClient(Endpoint endpoint, String name) {
    return endpoint.take(name, new ZiplineScope(), EchoServiceAdapter.INSTANCE);
  }

  /** Simulate generated code for inbound calls. */
  public static void bindEchoService(Endpoint endpoint, String name, EchoService echoService) {
    endpoint.bind(name, echoService, EchoServiceAdapter.INSTANCE);
  }

  /** Simulate generated code for outbound calls. */
  public static GenericEchoService<String> takeGenericEchoService(Endpoint endpoint, String name) {
    return endpoint.take(name, new ZiplineScope(), GenericEchoServiceAdapter.INSTANCE);
  }

  /** Simulate generated code for inbound calls. */
  public static void bindGenericEchoService(
      Endpoint endpoint, String name, GenericEchoService<String> echoService) {
    endpoint.bind(name, echoService, GenericEchoServiceAdapter.INSTANCE);
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter. */
  public static class EchoServiceAdapter extends ZiplineServiceAdapter<EchoService> {
    public static final EchoServiceAdapter INSTANCE = new EchoServiceAdapter();

    @Override public String getSimpleName() {
      return "EchoService";
    }

    @Override public String getSerialName() {
      return "app.cash.zipline.kotlin.EchoService";
    }

    @Override public List<KSerializer<?>> getSerializers() {
      return Collections.emptyList();
    }

    @Override public List<ZiplineFunction<EchoService>> ziplineFunctions(
        SerializersModule serializersModule) {
      KSerializer<?> requestSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, echoRequestKt);
      KSerializer<?> responseSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, echoResponseKt);
      String name = "fun echo(app.cash.zipline.testing.EchoRequest): app.cash.zipline.testing.EchoResponse";
      return Arrays.asList(
        new ReturningZiplineFunction<EchoService>(SignatureHashKt.signatureHash(name), name,
            Arrays.asList(requestSerializer), responseSerializer) {
          @Override
          public Object call(EchoService service, List<?> args) {
            return service.echo((EchoRequest) args.get(0));
          }
        }
      );
    }

    @Override public EchoService outboundService(OutboundCallHandler callHandler) {
      return new GeneratedOutboundService(callHandler);
    }

    private static class GeneratedOutboundService
        implements EchoService, OutboundService {
      private final OutboundCallHandler callHandler;

      GeneratedOutboundService(OutboundCallHandler callHandler) {
        this.callHandler = callHandler;
      }

      @Override public OutboundCallHandler getCallHandler() {
        return callHandler;
      }

      @Override public EchoResponse echo(EchoRequest request) {
        return (EchoResponse) callHandler.call(this, 0, request);
      }

      @Override public void close() {
      }
    }
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter. */
  public static class GenericEchoServiceAdapter
      extends ZiplineServiceAdapter<GenericEchoService<String>> {
    public static final GenericEchoServiceAdapter INSTANCE = new GenericEchoServiceAdapter();

    @Override public String getSimpleName() {
      return "GenericEchoService";
    }

    @Override public String getSerialName() {
      return "app.cash.zipline.kotlin.GenericEchoService<kotlin.String>";
    }

    @Override public List<KSerializer<?>> getSerializers() {
      return Collections.emptyList();
    }

    @Override
    public List<ZiplineFunction<GenericEchoService<String>>> ziplineFunctions(
        SerializersModule serializersModule) {
      KSerializer<?> requestSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, stringKt);
      KSerializer<?> responseSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, listOfStringKt);
      String name = "fun genericEcho(T): kotlin.collections.List<T>";
      return Arrays.asList(
        new ReturningZiplineFunction<GenericEchoService<String>>(
            SignatureHashKt.signatureHash(name), name, Arrays.asList(requestSerializer),
            responseSerializer) {
          @Override public Object call(GenericEchoService<String> service, List<?> args) {
            return service.genericEcho((String) args.get(0));
          }
        });
    }

    @Override public GenericEchoService<String> outboundService(OutboundCallHandler callHandler) {
      return new GeneratedOutboundService(callHandler);
    }

    private static class GeneratedOutboundService
        implements GenericEchoService<String>, OutboundService {
      private final OutboundCallHandler callHandler;

      GeneratedOutboundService(OutboundCallHandler callHandler) {
        this.callHandler = callHandler;
      }

      @Override public OutboundCallHandler getCallHandler() {
        return callHandler;
      }

      @Override public List<String> genericEcho(String request) {
        return (List<String>) callHandler.call(this, 0, request);
      }

      @Override public void close() {
      }
    }
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter. */
  public static class EchoZiplineServiceAdapter extends ZiplineServiceAdapter<EchoZiplineService> {
    public static final EchoZiplineServiceAdapter INSTANCE = new EchoZiplineServiceAdapter();

    @Override public String getSimpleName() {
      return "EchoService";
    }

    @Override public String getSerialName() {
      return "app.cash.zipline.kotlin.EchoZiplineService";
    }

    @Override public List<KSerializer<?>> getSerializers() {
      return Collections.emptyList();
    }

    @Override
    public List<ZiplineFunction<EchoZiplineService>> ziplineFunctions(
        SerializersModule serializersModule) {
      KSerializer<?> requestSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, echoRequestKt);
      KSerializer<?> responseSerializer
        = (KSerializer) SerializersKt.serializer(serializersModule, echoResponseKt);
      String name = "fun echo(app.cash.zipline.testing.EchoRequest): app.cash.zipline.testing.EchoResponse";
      return Arrays.asList(
        new ReturningZiplineFunction<EchoZiplineService>(SignatureHashKt.signatureHash(name), name,
            Arrays.asList(requestSerializer), responseSerializer) {
          @Override public Object call(EchoZiplineService service, List<?> args) {
            return service.echo((EchoRequest) args.get(0));
          }
        });
    }

    @Override public EchoZiplineService outboundService(OutboundCallHandler callHandler) {
      return new GeneratedOutboundService(callHandler);
    }

    private static class GeneratedOutboundService
        implements EchoZiplineService, OutboundService {
      private final OutboundCallHandler callHandler;

      GeneratedOutboundService(OutboundCallHandler callHandler) {
        this.callHandler = callHandler;
      }

      @Override public OutboundCallHandler getCallHandler() {
        return callHandler;
      }

      @Override public EchoResponse echo(EchoRequest request) {
        return (EchoResponse) callHandler.call(this, 0, request);
      }

      @Override public void close() {
      }
    }
  }

  /** Simulate generated code for inbound calls. */
  public static void bindEchoZiplineService(
      Endpoint endpoint, String name, EchoZiplineService service) {
    endpoint.bind(name, service, EchoZiplineServiceAdapter.INSTANCE);
  }

  /** Simulate generated code for outbound calls. */
  public static EchoZiplineService takeEchoZiplineService(Endpoint endpoint, String name) {
    return endpoint.take(name, new ZiplineScope(), EchoZiplineServiceAdapter.INSTANCE);
  }

  private ZiplineTestInternals() {
  }
}

File: zipline-testing/src/hostMain/kotlin/app/cash/zipline/testing/ZiplineTestInternals.kt
/*
 * Copyright (C) 2023 Cash App
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress(
  // Access :zipline internals
  "CANNOT_OVERRIDE_INVISIBLE_MEMBER",
  "EXPOSED_FUNCTION_RETURN_TYPE",
  "EXPOSED_PARAMETER_TYPE",
  "EXPOSED_SUPER_CLASS",
  "INVISIBLE_MEMBER",
  "INVISIBLE_REFERENCE",
)

package app.cash.zipline.testing

import app.cash.zipline.ZiplineFunction
import app.cash.zipline.ZiplineScope
import app.cash.zipline.internal.bridge.Endpoint
import app.cash.zipline.internal.bridge.OutboundCallHandler
import app.cash.zipline.internal.bridge.OutboundService
import app.cash.zipline.internal.bridge.ReturningZiplineFunction
import app.cash.zipline.internal.bridge.ZiplineServiceAdapter
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.reflect.typeOf
import kotlinx.coroutines.CoroutineScope
import kotlinx.serialization.KSerializer
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.serializer

/**
 * Suppress visibility warnings to directly call internal functions annotated [PublishedApi].
 *
 * Note that when overriding functions from invisible members, this class uses the base generic type
 * because that interferes the least with our suppressed warnings.
 */
object ZiplineTestInternals {
  private val echoResponseKt = typeOf<EchoResponse>()
  private val echoRequestKt = typeOf<EchoRequest>()
  private val stringKt = typeOf<String>()
  private val listOfStringKt = typeOf<List<String>>()

  fun newEndpointPair(): Pair<Endpoint, Endpoint> {
    val scope = CoroutineScope(EmptyCoroutineContext)
    return newEndpointPair(scope, kotlinBuiltInSerializersModule)
  }

  /** Simulate generated code for outbound calls.  */
  fun takeEchoClient(endpoint: Endpoint, name: String): EchoService {
    return endpoint.take(name, ZiplineScope(), EchoServiceAdapter())
  }

  /** Simulate generated code for inbound calls.  */
  fun bindEchoService(endpoint: Endpoint, name: String, echoService: EchoService) {
    endpoint.bind(name, echoService, EchoServiceAdapter())
  }

  /** Simulate generated code for outbound calls.  */
  fun takeGenericEchoService(endpoint: Endpoint, name: String): GenericEchoService<String> {
    return endpoint.take(name, ZiplineScope(), GenericEchoServiceAdapter())
  }

  /** Simulate generated code for inbound calls.  */
  fun bindGenericEchoService(
    endpoint: Endpoint,
    name: String,
    echoService: GenericEchoService<String>,
  ) {
    endpoint.bind(name, echoService, GenericEchoServiceAdapter())
  }

  /** Simulate generated code for inbound calls.  */
  fun bindEchoZiplineService(
    endpoint: Endpoint,
    name: String,
    service: EchoZiplineService,
  ) {
    endpoint.bind(name, service, EchoZiplineServiceAdapter())
  }

  /** Simulate generated code for outbound calls.  */
  fun takeEchoZiplineService(endpoint: Endpoint, name: String): EchoZiplineService {
    return endpoint.take(name, ZiplineScope(), EchoZiplineServiceAdapter())
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter.  */
  class EchoServiceAdapter : ZiplineServiceAdapter<EchoService>() {
    override val simpleName: String
      get() = "EchoService"
    override val serialName: String
      get() = "app.cash.zipline.kotlin.EchoService"
    override val serializers: List<KSerializer<*>>
      get() = emptyList()

    override fun ziplineFunctions(
      serializersModule: SerializersModule,
    ): List<ZiplineFunction<EchoService>> {
      val requestSerializer = serializersModule.serializer(echoRequestKt) as KSerializer<*>
      val responseSerializer = serializersModule.serializer(echoResponseKt) as KSerializer<*>
      val name =
        "fun echo(app.cash.zipline.testing.EchoRequest): app.cash.zipline.testing.EchoResponse"
      return listOf<ZiplineFunction<EchoService>>(
        object : ReturningZiplineFunction<EchoService>(
          name.signatureHash(),
          name,
          listOf(requestSerializer),
          responseSerializer,
        ) {
          override fun call(service: EchoService, args: List<*>): Any {
            return service.echo(args[0] as EchoRequest)
          }
        },
      )
    }

    override fun outboundService(callHandler: OutboundCallHandler): EchoService {
      return GeneratedOutboundService(callHandler)
    }

    private class GeneratedOutboundService(
      override val callHandler: OutboundCallHandler,
    ) : EchoService,
      OutboundService {

      override fun echo(request: EchoRequest): EchoResponse {
        return callHandler.call(this, 0, request) as EchoResponse
      }
    }
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter.  */
  class GenericEchoServiceAdapter : ZiplineServiceAdapter<GenericEchoService<String>>() {
    override val simpleName: String
      get() = "GenericEchoService"
    override val serialName: String
      get() = "app.cash.zipline.kotlin.GenericEchoService<kotlin.String>"
    override val serializers: List<KSerializer<*>>
      get() = emptyList()

    override fun ziplineFunctions(
      serializersModule: SerializersModule,
    ): List<ZiplineFunction<GenericEchoService<String>>> {
      val requestSerializer = serializersModule.serializer(stringKt) as KSerializer<*>
      val responseSerializer = serializersModule.serializer(listOfStringKt) as KSerializer<*>
      val name = "fun genericEcho(T): kotlin.collections.List<T>"
      return listOf<ZiplineFunction<GenericEchoService<String>>>(
        object : ReturningZiplineFunction<GenericEchoService<String>>(
          name.signatureHash(),
          name,
          listOf(requestSerializer),
          responseSerializer,
        ) {
          override fun call(service: GenericEchoService<String>, args: List<*>): Any {
            return service.genericEcho(args[0] as String)
          }
        },
      )
    }

    override fun outboundService(callHandler: OutboundCallHandler): GenericEchoService<String> {
      return GeneratedOutboundService(callHandler)
    }

    private class GeneratedOutboundService(override val callHandler: OutboundCallHandler) :
      GenericEchoService<String>,
      OutboundService {

      override fun genericEcho(request: String): List<String> {
        return callHandler.call(this, 0, request) as List<String>
      }
    }
  }

  /** Simulate a generated subclass of ZiplineServiceAdapter.  */
  class EchoZiplineServiceAdapter : ZiplineServiceAdapter<EchoZiplineService>() {
    override val simpleName: String
      get() = "EchoService"
    override val serialName: String
      get() = "app.cash.zipline.kotlin.EchoZiplineService"
    override val serializers: List<KSerializer<*>>
      get() = emptyList()

    override fun ziplineFunctions(
      serializersModule: SerializersModule,
    ): List<ZiplineFunction<EchoZiplineService>> {
      val requestSerializer = serializersModule.serializer(echoRequestKt) as KSerializer<*>
      val responseSerializer = serializersModule.serializer(echoResponseKt) as KSerializer<*>
      val name =
        "fun echo(app.cash.zipline.testing.EchoRequest): app.cash.zipline.testing.EchoResponse"
      return listOf<ZiplineFunction<EchoZiplineService>>(
        object : ReturningZiplineFunction<EchoZiplineService>(
          name.signatureHash(),
          name,
          listOf(requestSerializer),
          responseSerializer,
        ) {
          override fun call(service: EchoZiplineService, args: List<*>): Any {
            return service.echo(args[0] as EchoRequest)
          }
        },
      )
    }

    override fun outboundService(callHandler: OutboundCallHandler): EchoZiplineService {
      return GeneratedOutboundService(callHandler)
    }

    private class GeneratedOutboundService(override val callHandler: OutboundCallHandler) :
      EchoZiplineService,
      OutboundService {

      override fun echo(request: EchoRequest): EchoResponse {
        return callHandler.call(this, 0, request) as EchoResponse
      }
    }
  }
}
